---
title: "Dasatinib Pathway Analysis"
author: "Kieran Redpath"
date: "23 March 2025"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 3
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Summary
* __Setup:__
+ Loads packages.
+ Loads and processes data from the CCLE (which is then normalised) and the GDSC.
+ Filters this down to just the relevant information, incorporating information on CDH1 mutation status.

* __Differential Expression Analysis:__
+ Carries out differential expression analysis for IC50 and AUC (individually), to determine what genes are significantly differentially expressed in dasatinib resistant/susceptible cell lines.
+ Finds the intersect between the best results for IC50 and AUC, determining that the relationship between them is important (even though results may not appear significant).
+ Saves this table as a .csv.

* __Processing IC50 and AUC Differential Expression Data:__
+ Creates new objects to be fed into pathway analyses.
+ Ranks samples based on average adjusted P value between IC50 and AUC.
+ Plots the relationship between IC50/AUC and gene expression data.

* __Pathway Analysis:__
+ Finds pathways based on a list of genes, using the "goseq" package (both length-corrected and non length-corrected).
+ Extracts genes in the dataset from each pathway.

* __Heatmap Construction:__
+ Builds heatmaps for all the genes in each pathway, and saves them as .png's.
+ Creates a heatmap of the genes shared between pathways.
+ Graphs the CDH1 mutation status of each cell line, showing the split between CDH1+ and CDH1- cell lines.

* __Print the Significant Pathways:__
+ Prints heatmaps in the document for the pathways with an adjusted p value of <0.05.

## Setup
### Load packages
```{r setup, results='hide', warning=FALSE, message=FALSE}
library(limma)
library(edgeR)
library(data.table)
library(magrittr)
library(ggplot2)
library(CePa)
library(tidyverse)
library(ggbeeswarm)
library(dplyr)
library(ReactomePA)
library(org.Hs.eg.db)
library(reactome.db)
library(goseq)
library(gplots)
source('Data/heatmap-mik.R')
library(knitr)
library(kableExtra)
library(devtools)
library(viridis)
library(openxlsx)
set.seed(42)
```

### Load CCLE gene expression data
```{r, results='hide'}
# Load Homo sapiens gene data (including Entrez ID's).
hs <- org.Hs.eg.db

# Load expression data from the CCLE, available here: https://depmap.org/portal/data_page/?tab=allData
CCLE <- read.gct('Data/CCLE_RNAseq_genes_counts_20180929.gct')
# Remove X's introduced by R for cell lines that start with a number, but not those that don't.
colnames(CCLE) <- sub("^X(\\d+)", "\\1", colnames(CCLE))

# Normalise CCLE counts using the limma voom method.
dge <- DGEList(counts = CCLE)
dge <- calcNormFactors(dge)
voom_dge <- voom(dge, plot = TRUE)
voom_CCLE <- voom_dge$E

# Tidy up the environment.
rm(list = c("dge", "voom_dge"))
```

### Load GDSC2 drug response data
```{r}
# Load the GDSC dataset, current release available here: www.cancerrxgene.org/downloads/bulk_download
# Archived releases available via ftp: ftp.sanger.ac.uk/project/cancerrxgene/releases/
GDSC2 <- fread("Data/GDSC2_fitted_dose_response_25Feb20.csv", sep= ';')

# Load the GDSC dataset, current release available here: www.cancerrxgene.org/downloads/bulk_download
# Archived releases available via ftp: ftp.sanger.ac.uk/project/cancerrxgene/releases/
# GDSC2 <- read.xlsx("Data/GDSC2_fitted_dose_response_27Oct23.xlsx")

# Filter for cell lines that have data for the drug "Dasatinib".
dasatinib_GDSC2 <- filter(GDSC2, DRUG_NAME == "Dasatinib")
```

# Format data
```{r}
# Standardize CCLE and GDSC2 cell line names, resolving a naming discrepancy between between cellosaurus cell lines www.cellosaurus.org/CVCL_3174 and www.cellosaurus.org/CVCL_1774.
# CCLE:
colnames(voom_CCLE)[colnames(voom_CCLE) == "TT_OESOPHAGUS"] <- "T-T_OESOPHAGUS"
# Isolate cell line and tissue labels for later indexing.
split_cell_line_tissues <- colnames(voom_CCLE) %>% strsplit(., "_")
colnames(voom_CCLE) <- strsplit(colnames(voom_CCLE), "_") %>% 
  lapply(., function(x) x[1]) %>% 
  unlist()
# GDSC2:
dasatinib_GDSC2$CELL_LINE_NAME <- ifelse(
  dasatinib_GDSC2$CELL_LINE_NAME == "T-T", "T-T", # if
  gsub("-", "", dasatinib_GDSC2$CELL_LINE_NAME, fixed = TRUE) # else
)

# Filter datasets for common cell lines.
common_lines <- intersect(dasatinib_GDSC2$CELL_LINE_NAME,colnames(voom_CCLE))
match_voom_CCLE <- match(common_lines, colnames(voom_CCLE))
common_voom_CCLE <- voom_CCLE[ , na.omit(match_voom_CCLE)]
match_dasatinib_GDSC2 <- match(common_lines, dasatinib_GDSC2$CELL_LINE_NAME)
common_dasatinib_GDSC2 <- dasatinib_GDSC2[na.omit(match_dasatinib_GDSC2),]
colnames(common_dasatinib_GDSC2)[5] <- "CCLE_Name"

# Tidy up the environment.
rm(list = c("common_lines", "match_voom_CCLE", "match_dasatinib_GDSC2"))
```

# Filter data
```{r}
# Filter CCLE and GDSC2 data for BRCA and STAD cell lines with data for dasatinib.
# GDSC (also adds a column for CDH1 expression data):
sort_dasatinib_GDSC2 <- common_dasatinib_GDSC2 %>%
  filter(TCGA_DESC %in% c("BRCA", "STAD")) %>%
  filter(AUC < quantile(AUC, 0.33) | AUC > quantile(AUC, 0.66))
sort_dasatinib_GDSC2$CDH1 <- common_voom_CCLE %>% as.data.frame() %>%
  dplyr::select(any_of(sort_dasatinib_GDSC2$CCLE_Name)) %>%
  dplyr::slice(match("ENSG00000039068.14", rownames(.))) %>%
  as.vector() %>%
  t()

# CCLE:
sort_voom_CCLE <- common_voom_CCLE %>%
  as.data.frame() %>%
  dplyr::select(any_of(sort_dasatinib_GDSC2$CCLE_Name)) %>% 
  as.matrix()

# Tidy up the environment.
rm(list = c("common_voom_CCLE", "common_dasatinib_GDSC2"))
```

# Index cell lines to tissues
```{r}
# Create a tool to find tissue type based on cell line name or vice versa.
tissue_tool <- data.frame(
  Cell_Line = sapply(split_cell_line_tissues, `[[`, 1),
  Tissue_Type = sapply(split_cell_line_tissues, function(x) paste(x[-1], collapse = "_")),
  stringsAsFactors = FALSE)
tissue_tool <- tissue_tool %>% mutate(
    Tissue_Type = Tissue_Type %>%
      str_remove_all("c") %>%
      str_replace_all("[[:punct:]]", "_"))

# Create another tool for the same purpose, but only including cell lines of interest.
sort_tissue_tool <- tissue_tool %>% filter(Tissue_Type %in% c("BREAST", "STOMACH"))
common_lines <- intersect(colnames(sort_voom_CCLE), sort_tissue_tool$Cell_Line)
sort_tissue_tool <- sort_tissue_tool %>%
  filter(Cell_Line %in% common_lines) %>%
  arrange(factor(Cell_Line, levels = common_lines))


# Tidy up the environment.
rm(list = c("split_cell_line_tissues", "common_lines"))

# Remove a cell line that causes problems: 2313287
# Remove column from sort_voom_CCLE
# sort_voom_CCLE <- sort_voom_CCLE[,!colnames(sort_voom_CCLE) %in% "2313287"]
# # Have to remove this from other objects
# sort_dasatinib_GDSC2 <- sort_dasatinib_GDSC2[sort_dasatinib_GDSC2$CCLE_Name != "2313287",]
# tissuetoolsort <- tissuetoolsort[tissuetoolsort$Cell_Line != "2313287",]
```

## Differential Expression Analysis

### Differential expression analysis for IC50
```{r}
# Prints some basic information about the objects that will be used for further analyses, for reference.
sum(colnames(sort_voom_CCLE)==sort_dasatinib_GDSC2$CCLE_Name)
names(sort_dasatinib_GDSC2)

# Fits the expression data to a linear model. 
group <- ifelse(sort_dasatinib_GDSC2$LN_IC50 > median(sort_dasatinib_GDSC2$LN_IC50), "High", "Low")
table(group)
boxplot(sort_dasatinib_GDSC2$LN_IC50 ~ group)
design = model.matrix(~group);
design %>% head()
colnames(design) = c("Mean", "HighVsLow")
fit = lmFit(sort_voom_CCLE, design)
fit = eBayes(fit)
tt = topTable(fit, coef="HighVsLow", adjust="BH",n=nrow(sort_voom_CCLE))
options(digits=4)

# Plots the first gene on the list, and counts the number of significant samples (threshold of 0.01).
plot(density(sort_voom_CCLE[,1]))
sum(tt$adj.P.Val<0.05)

# Creates a volcano plot of the IC50 fold change vs. adjusted P value, showing significant samples with high fold change in red.
sigFC = (tt$adj.P.Val < 0.05)  & (abs(tt$logFC) > 1)
volcanoplot(fit, coef="HighVsLow")
points(tt$logFC[which(sigFC)], 
       -log10(tt$P.Value[which(sigFC)]), 
       cex=0.6, col='red', pch=16)
abline(h = min(-log10(tt$P.Value[which(sigFC)])), lty=2, col='blue')
abline(v = c(-1,1), lty=2, col='blue')

# Provides basic information on the top table.
split <- strsplit(rownames(tt),".", fixed=T) %>% lapply(., function(x) x[1]) %>% unlist()
geneNames <- AnnotationDbi::select(org.Hs.eg.db, keys = split, column = c("SYMBOL","GENENAME"), key="ENSEMBL")
dim(tt)
tt$symbol <- geneNames$SYMBOL[match(split, geneNames$ENSEMBL)]
invisible(setDT(tt, keep.rownames = TRUE)[])

# Prints a list of the top 50 associated genes (can be altered to create larger lists for pathway analysis). Also creates a histogram and shows the median of IC50 values.
# cat(na.omit(tt$symbol[1:50]),sep="\n")
hist(sort_dasatinib_GDSC2$LN_IC50,50)
median(sort_dasatinib_GDSC2$LN_IC50)

# Finds the top expression values, visualising the positive correlation between gene expression and IC50.
topExp <- sort_voom_CCLE[match(tt$rn[1], rownames(sort_voom_CCLE)),]
df <- data.frame(topGene=topExp, ic50=group)
ggplot(df, aes(x=ic50, y=topGene)) + geom_boxplot()
```

### Differential expression analysis for AUC
```{r}
# Prints some basic information about the objects that will be used for further analyses, for reference.
sum(colnames(sort_voom_CCLE)==sort_dasatinib_GDSC2$CCLE_Name)
names(sort_dasatinib_GDSC2)

# Fits the expression data to a linear model.
group2 <- ifelse(sort_dasatinib_GDSC2$AUC > median(sort_dasatinib_GDSC2$AUC), "High", "Low")
table(group2)
boxplot(sort_dasatinib_GDSC2$AUC ~ group2)
design2 = model.matrix(~group2);
design2 %>% head()
colnames(design2) = c("Mean"
                      ,"HighVsLow"
)
fit2 = lmFit(sort_voom_CCLE, design2)
fit2 = eBayes(fit2)
tt2 = topTable(fit2, coef="HighVsLow", adjust="BH",n=nrow(sort_voom_CCLE))
options(digits=4)

# Plots the first gene on the list, and counts the number of significant samples (threshold of 0.05).
plot(density(sort_voom_CCLE[,1]))
sum(tt2$adj.P.Val<0.05)

# Creates a volcano plot of the AUC fold change vs. adjusted P value, showing significant samples with high fold change in red.
sigFC2 = (tt2$adj.P.Val < 0.05)  & (abs(tt2$logFC) > 1)
volcanoplot(fit2, coef="HighVsLow")
points(tt2$logFC[which(sigFC2)], 
       -log10(tt2$P.Value[which(sigFC2)]), 
       cex=0.6, col='red', pch=16)
abline(h = min(-log10(tt2$P.Value[which(sigFC2)])), lty=2, col='blue')
abline(v = c(-1,1), lty=2, col='blue')

# Provides basic information on the top table, and prints the first 20 genes, for reference.
split2 <- strsplit(rownames(tt2),".", fixed=T) %>% lapply(., function(x) x[1]) %>% unlist()
geneNames2 <- AnnotationDbi::select(org.Hs.eg.db, keys = split2, column = c("SYMBOL","GENENAME"), key="ENSEMBL")
dim(tt2)
tt2$symbol <- geneNames2$SYMBOL[match(split2, geneNames2$ENSEMBL)]
invisible(setDT(tt2, keep.rownames = TRUE)[])
tt2[1:20,] %>% kable() %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=8), fixed_thead=T)

# Prints a list of the top 50 associated genes (can be altered to create larger lists for pathway analysis). Also creates a histogram and shows the median of AUC values.
cat(na.omit(tt2$symbol[1:50]),sep="\n")
hist(sort_dasatinib_GDSC2$AUC,50)
median(sort_dasatinib_GDSC2$AUC)

# Finds the top expression values, visualising the negative correlation between gene expression and AUC.
topExp2 <- sort_voom_CCLE[match(tt2$rn[1], rownames(sort_voom_CCLE)),]
df2 <- data.frame(topGene=topExp2, AUC=group)
ggplot(df2, aes(x=AUC, y=topGene)) + geom_boxplot()
```

## Processing IC50 and AUC Differential Expression Data
### Finding the overlap between significant samples
* This information (present in top table 3) will be used for pathway analysis.
```{r}
# Creates a new top table, containing the information for both IC50 and AUC, with ensembl ID's and gene symbols.
tt3 <- full_join(tt, tt2, by= "rn")
tt3 <- dplyr::select(tt3, -c("symbol.x", "P.Value.y", "P.Value.x", "t.y", "t.x", "B.x", "B.y", "AveExpr.x"))
tt3 <- dplyr::rename(tt3, "Gene_Symbol" = "symbol.y", "AUC_logFC" = "logFC.y", "Avg_Exp" = "AveExpr.y", "AUC_Adj_PVal" = "adj.P.Val.y", "IC50_logFC" = "logFC.x", "IC50_Adj_PVal" = "adj.P.Val.x", "Ensembl_ID" = "rn")
tt3 <- tt3[ ,c(7,1,5,2,3,4,6)]

# Identifies that the sign of the log fold change is consistent with the IC50 and AUC relationship. As IC50 and AUC have an opposite association with cell line resistance/susceptibility to dasatinib, only samples with -log fold change signs (IC50 log FC sign*AUC log FC sign) are kept. Need to check this.
tt3 <- mutate(tt3, sign(tt3$IC50_logFC), sign(tt3$AUC_logFC))
tt3 <- mutate(tt3, sign(tt3$IC50_logFC)*sign(tt3$AUC_logFC))
tt3 <- dplyr::rename(tt3, "logFC_Sign" = "sign(tt3$IC50_logFC) * sign(tt3$AUC_logFC)", "IC50_Sign" = "sign(tt3$IC50_logFC)", "AUC_Sign" = "sign(tt3$AUC_logFC)")

# Ranks samples based on average rank between IC50 and AUC adjusted P values.
tt3 <-
  tt3 %>% mutate(., rank_ic50=rank(IC50_Adj_PVal)) %>%
  mutate(., rank_auc=rank(AUC_Adj_PVal)) %>%
  mutate(., avg_rank=0.5*(rank_auc + rank_ic50)) %>%
  arrange(., avg_rank)
```

### Create plots illustrating the relationship between AUC/IC50 and gene expression
* Helps visualise the relationship between these variables. Tells you what a positive vs negative log fold change means
```{r}
# Creates a boxplot and scatter plot of gene expression vs IC50.
ggplot(data=tt3, mapping=aes(x=IC50_Sign,y=Avg_Exp, group=IC50_Sign, fill=IC50_Sign)) +
  geom_boxplot()
ggplot(data=tt3, mapping=aes(x=IC50_logFC,y=Avg_Exp)) +geom_point()

# Creates a boxplot and scatter plot of gene expression vs AUC.
ggplot(data=tt3, mapping=aes(x=AUC_Sign,y=Avg_Exp, group=AUC_Sign, fill=AUC_Sign)) +
  geom_boxplot()
ggplot(data=tt3, mapping=aes(x=AUC_logFC,y=Avg_Exp)) +geom_point()
```

## Pathway Analysis
### GoSeq analysis with the top genes from AUC and IC50
* "goseq" corrects for gene length, so can be a more effective tool than "ReactomePA".
```{r, warning=FALSE, message=FALSE}
# Defines the significant samples. This is based on positive log fold change sign (IC50 sign*AUC sign, to check if directionality per sample is the same for both values), and an absolute log fold change of log2(2) for both IC50 and AUC. Ranks samples based on adjusted p value, averaged between IC50 and AUC.
SigSamples <- filter(tt3, tt3$logFC_Sign == 1, abs(tt3$IC50_logFC) > log2(2), abs(tt3$AUC_logFC) > log2(2))
SigSamples <- 
  SigSamples %>% mutate(., rank_ic50=rank(IC50_Adj_PVal)) %>%
  mutate(., rank_auc=rank(AUC_Adj_PVal)) %>% 
  mutate(., avg_rank=0.5*(rank_auc + rank_ic50)) %>% 
  arrange(., avg_rank)

# Cuts down to the top significant samples, for pathway analysis, and adds Entrez ID's to these.
# Test using all samples
TopXGenes <- as.character(na.omit(SigSamples$Gene_Symbol))
TopXGenesEntrez <- AnnotationDbi::select(hs, 
                                         keys = TopXGenes,
                                         columns = c("ENTREZID", "SYMBOL"),
                                         keytype = "SYMBOL")

# Adds Entrez ID's to and define non-significant samples.
tt3Genes <- as.character(na.omit(tt3$Gene_Symbol))
tt3GenesEntrez <- AnnotationDbi::select(hs, 
                                        keys = tt3Genes,
                                        columns = c("ENTREZID", "SYMBOL"),
                                        keytype = "SYMBOL")

# Gets pathway names and extracts human-only pathways. Also defines the intersect of the significant genes and those that are in reactome.
rName <- as.list(reactomePATHNAME2ID)
rName <- rName[grep("Homo sapiens", names(rName))]
rGenes <- as.list(reactomePATHID2EXTID)
rGenesPath <- rGenes[match(rName, names(rGenes))]
rGenesPath <- lapply(rGenesPath, unique)
rGeneByPath <- as.list(reactomeEXTID2PATHID)
allGenes <- intersect( tt3GenesEntrez$ENTREZID, unique(unlist(rGenesPath)) )
length(allGenes)
sigGenes <- intersect( TopXGenesEntrez$ENTREZID, unique(unlist(rGenesPath)) )
length(sigGenes)
plotGenes <- rep(0, length(allGenes))
names(plotGenes) <- allGenes
plotGenes[match(sigGenes, names(plotGenes))] <- 1

# Counts the number of significant genes (1's) and non-significant genes (0's). Finds the genes relevant to your dataset.
table(plotGenes)
mt <- match(allGenes, names(rGeneByPath))
rGeneByPath <- lapply(rGeneByPath[mt], function(x) intersect(x, names(rGenesPath)))

# Carries out the goseq analysis: length-corrected, then not length-corrected. The venn diagram shows that most of the pathways are the same between the two. 
pwf <- nullp(plotGenes, 'hg19', id = "knownGene", plot.fit = TRUE)
goseqReactome <- goseq(pwf, gene2cat = rGeneByPath)
goseqReactome$adjP <- p.adjust(goseqReactome$over_represented_pvalue, method = "fdr")

# Filters the length-corrected pathways for significance (threshold of 0.05) and adds pathway information.
goseqPathways <- filter(goseqReactome, goseqReactome$adjP <1)
rPathName <- as.list(reactomePATHID2NAME)
goseqPathways$Pathway <- gsub("Homo sapiens: ", "", rPathName[match(goseqPathways$category, names(rPathName))])
SiggoseqPathways <- goseqPathways
```

### Extract genes from GoSeq pathway analysis
* Adds more information to "goseqPathways", so we can directly identify the genes involved.
* Creates a list of significant genes for each goseq pathway, converts them to symbol ID's. Also does the same for all the genes in each pathway and adds the information to "goseqPathways". i.e. siggoseqpathways = overrepresented genes in paths (numDEInCat), goseqpathways = all genes in paths (numInCat)
```{r}
# Creates a tool to easily match genes, between gene symbol, entrez ID, and ensembl ID.
GeneLabelTool <- dplyr::pull(tt3, Gene_Symbol)
GeneLabelTool <- AnnotationDbi::select(hs,
                                       keys = GeneLabelTool,
                                       columns = c("ENSEMBL", "ENTREZID", "SYMBOL"),
                                       keytype = "SYMBOL")

# Creates a list of significant genes for each goseq pathway, converts them to symbol ID's. Also does the same for all the genes in each pathway and adds the information to "goseqPathways".
SiggenesinPaths <-list()
for(i in 1:nrow(goseqPathways)){
  SiggenesinPaths[[i]] <- rGenesPath[match(goseqPathways$category[i], names(rGenesPath))] %>% 
    .[[1]] %>% 
    intersect(., rownames(pwf)[pwf$DEgenes==1])
}

genesinPaths <- list()
for(i in 1:nrow(goseqPathways)){
  genesinPaths[[i]] <- rGenesPath[match(goseqPathways$category[i], names(rGenesPath))] %>% 
    .[[1]] %>% 
    intersect(., rownames(pwf)[pwf$DEgenes==1 | pwf$DEgenes==0])
}
SigsymbolsinPaths <- lapply(SiggenesinPaths, function(x) GeneLabelTool$SYMBOL[na.omit(match(x, GeneLabelTool$ENTREZID))] )
symbolsinPaths <- lapply(genesinPaths, function(x) GeneLabelTool$SYMBOL[na.omit(match(x, GeneLabelTool$ENTREZID))] )

# Save the object that's loaded in to calculate metagenes.
ensginPaths <- lapply(genesinPaths, function(x) GeneLabelTool$ENSEMBL[na.omit(match(x, GeneLabelTool$ENTREZID))] )
saveRDS(ensginPaths, file = "Data/Ensembl_ID_in_Paths.Rds")

genesStick <- lapply(symbolsinPaths, function(x) paste0(x, collapse="::", sep="")) %>% unlist()
goseqPathways$DEgenesInCat <- genesStick
goseqPathways <- lapply(goseqPathways, gsub, pattern='/', replacement=' ') %>% as.data.frame()
goseqPathways$Pathway <- lapply(goseqPathways$Pathway, gsub, pattern=':', replacement='-')

# Gives a table with which to identify the genes involved in each pathway, and saves it - despite the name, siggoseq pathways does not filter for significance, it simply makes it easier to plot data.
SiggenesStick <- lapply(SigsymbolsinPaths, function(x) paste0(x, collapse="::", sep="")) %>% unlist()
SiggoseqPathways$DEgenesInCat <- SiggenesStick
SiggoseqPathways <- lapply(SiggoseqPathways, gsub, pattern='/', replacement=' ') %>% as.data.frame()
SiggoseqPathways$Pathway <- lapply(SiggoseqPathways$Pathway, gsub, pattern=':', replacement='-')
# SiggoseqPathways <- lapply(SiggoseqPathways$Pathway, gsub, pattern='ERBB2:ERBB3', replacement='ERBB2-ERBB3') %>% as.data.frame()
SiggoseqPathways$Pathway <- as.character(SiggoseqPathways$Pathway)
write.csv(SiggoseqPathways, file = "Data/goseq_Das_Pathways.csv")
```

## Heatmap  Construction
### Setup for expression data
* Creates objects that add more information to the heatmap. Also prints a legend, showing which cell type is associated with which colour on the heatmap.
```{r}
# Adds more information to "tissuetoolsort". This tells us which group each cell line belongs to in terms of IC50 (High/Low -log IC50) and AUC (High/Low AUC) response, and CDH1 expression level in these cell lines. These are defined as the top/bottom 50% of the data for AUC/-log IC50, and CDH1- Vs. CDH1+, and prints "tissuetoolsort" inline.
tissuetoolsort$IC50_Group <- ifelse(sort_dasatinib_GDSC2$LN_IC50 > median(sort_dasatinib_GDSC2$LN_IC50), "High", "Low")
tissuetoolsort$AUC_Group <- ifelse(sort_dasatinib_GDSC2$AUC > median(sort_dasatinib_GDSC2$AUC), "High", "Low")
tissuetoolsort$CDH1_Group <- ifelse(sort_dasatinib_GDSC2$CDH1 > 2, "High", "Low")
tissuetoolsort$CDH1_Level <- sort_dasatinib_GDSC2$CDH1
tissuetoolsort %>% kable() %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=8), fixed_thead=T)

# Creates a color table for the heatmap column bars, which lets us sort by tissue type, AUC, -log IC50, or CDH1.
cc <- rbind((as.factor(tissuetoolsort$Tissue_Type) %>% as.numeric() %>% rainbow(length(table(.)))[.]),
            c("darkblue", "blue")[as.numeric(as.factor(tissuetoolsort$IC50_Group))],
            c("darkblue", "blue")[as.numeric(as.factor(tissuetoolsort$AUC_Group))],
            # c("orange", "yellow")[as.numeric(as.factor(tissuetoolsort$CDH1_Group))],
            (as.factor(tissuetoolsort$CDH1_Group) %>% as.numeric() %>% cividis(length(table(.)))[.]),
            (as.factor(tissuetoolsort$CDH1_Level) %>% as.numeric() %>% cividis(length(table(.)), direction= -1)[.])
)
rownames(cc) <- c("Tissue Type", "-log IC50 Group", "AUC Group", "CDH1 Group", "CDH1 Level")

# Creates an object showing which gene is part of which pathway, for all genes and just the significant genes.
ensginPaths <- lapply(genesinPaths, function(x) GeneLabelTool$ENSEMBL[na.omit(match(x, GeneLabelTool$ENTREZID))] )
SigensginPaths <- lapply(SiggenesinPaths, function(x) GeneLabelTool$ENSEMBL[na.omit(match(x, GeneLabelTool$ENTREZID))] )

# Standardises names between "voom_CCLE" and "GeneLabelTool" for readability and so the code actually works.
rownames(sort_voom_CCLE) <- strsplit(rownames(sort_voom_CCLE),".", fixed=T) %>% lapply(., function(x) x[1]) %>% unlist()
# Save this to be read into the metagene analysis.
write.csv(sort_voom_CCLE, file = "Data/BRCA_STAD_Das_Expression_Data.csv")#, row.names = rownames(sort_voom_CCLE))

# Plots CDH1 expresssion level to identify CDH1- cell lines.
ggplot(sort_dasatinib_GDSC2, aes(x="", y=CDH1, colour=CDH1 > 2)) +
  ggtitle("CDH1 Expression Level in Cell Lines") + labs(y="Normalised CDH1 Expression", x="AUC") +
  scale_color_viridis(discrete=TRUE, option="cividis") +
  geom_beeswarm(show.legend = FALSE)

sort_dasatinib_GDSC2$AUC > quantile(sort_dasatinib_GDSC2$AUC, 0.5)
```

# Write out tt3 as a .csv for manipulating into a thesis table
```{r}
write.csv(tt3, file = "Thesis/tt3.csv")
```